src/zkevm_specs/state.py

```diff
@@ -0,0 +1,540 @@
+    get_tx_id = lambda row: row.keys[1]
+    get_addr = lambda row: row.keys[2]
+    get_storage_key = lambda row: row.keys[4]
+
+    # 0. Unused keys are 0
+    assert row.keys[3] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_account_destructed(row: Row, row_prev: Row):
+    get_addr = lambda row: row.keys[2]
+
+    # 0. Unused keys are 0
+    assert row.keys[1] == 0
+    assert row.keys[3] == 0
+    assert row.keys[4] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_state_row(row: Row, row_prev: Row, randomness: FQ):
+    #
+    # Constraints that affect all rows, no matter which Tag they use
+    #
+
+    # 0. Tags transition
+    #
+    # Tag monotonically increases for all values except for Start
+    if row.tag() != Tag.Start:
+        tag_diff = row.tag() - row_prev.tag()
+        assert_in_range(tag_diff, 0, MAX_KEY_DIFF)
+
+    # 1. key2 is linear combination of 10 x 16bit limbs
+    for limb in row.key2_limbs:
+        assert_in_range(limb, 0, 2**16 - 1)
+    assert row.keys[2] == linear_combine(row.key2_limbs, 2**16)
+
+    # 2. key4 is RLC encoded
+    for limb in row.key4_bytes:
+        assert_in_range(limb, 0, 2**8 - 1)
+    assert row.keys[4] == linear_combine(row.key4_bytes, randomness)
+
+    # 3. is_write is boolean
+    assert row.is_write in [0, 1]
+
+    # 4. Keys are sorted in lexicographic order for same Tag
+    #
+    # When in two consecutive rows the keys are equal in a column:
+    # - The corresponding keys in the following column must be increasing.
+    #
+    # key4 is RLC encoded, so it doesn't keep the order.  We compare each limb
+    # of key4 from highest byte to lowest byte.
+    def get_keys_in_order(row: Row) -> List[FQ]:
+        return (
+            [row.keys[0], row.keys[1]]
+            + [x for x in reversed(row.key2_limbs)]
+            + [row.keys[3]]
+            + [x for x in reversed(row.key4_bytes)]
+        )
+
+    keys = get_keys_in_order(row)
+    keys_prev = get_keys_in_order(row_prev)
+    keys_eq = True
+    for i in range(len(keys) - 1):
+        keys_eq = keys_eq and keys[i] == keys_prev[i]
+        if keys_eq:
+            key_diff = keys[i + 1] - keys_prev[i + 1]
+            assert_in_range(key_diff, 0, MAX_KEY_DIFF)
+
+    # 5. RWC is monotonically strictly increasing for a set of all keys
+    #
+    # When all the keys are equal in two consecutive a rows:
+    # - The corresponding rwc must be strictly increasing.
+    if all_keys_eq(row, row_prev):
+        rw_diff = row.rw_counter - row_prev.rw_counter
+        assert_in_range(rw_diff, 0, MAX_KEY_DIFF)
+
+    # 6. Read consistency
+    #
+    # When a row is READ
+    # AND When all the keys are equal in two consecutive a rows:
+    # - The corresponding value must be equal to the previous row
+    if row.is_write == 0 and all_keys_eq(row, row_prev):
+        assert row.value == row_prev.value
+
+    # 7. Padding rows
+    # TODO: Figure out if we really need padding rows
```

Comment:
Hm, I forgot all the reasons for padding, but one of the problems without padding would probably be to ignore the constraints about the increasing counter because the address doesn't change in empty rows.

Reply:
From a discussion with Han, we can reuse the Start tag as padding (padding at the beginning instead of at the end)